src/inference/ess.jl

```diff
@@ -0,0 +1,141 @@
+"""
+    ESS
+
+Elliptical slice sampling algorithm.
+
+# Examples
+```jldoctest; setup = :(Random.seed!(1))
+julia> @model gdemo(x) = begin
+           m ~ Normal()
+           x ~ Normal(m, 0.5)
+       end
+gdemo (generic function with 2 methods)
+
+julia> sample(gdemo(1.0), ESS(), 1_000) |> mean
+Mean
+
+│ Row │ parameters │ mean     │
+│     │ Symbol     │ Float64  │
+├─────┼────────────┼──────────┤
+│ 1   │ m          │ 0.811555 │
+```
+"""
+struct ESS{space} <: InferenceAlgorithm end
+
+ESS() = ESS{()}()
+ESS(space::Symbol) = ESS{(space,)}()
+
+mutable struct ESSState{V<:VarInfo} <: AbstractSamplerState
+    vi::V
+end
+
+ESSState(model::Model) = ESSState(VarInfo(model))
+
+function Sampler(alg::ESS, model::Model, s::Selector)
+    # sanity check
+    space = getspace(alg)
+    if isempty(space)
+        pvars = get_pvars(model)
+        length(pvars) == 1 ||
+            error("[ESS] no symbol specified to sampler although there is not exactly one model parameter ($pvars)")
+    end
+
+    state = ESSState(model)
+    info = Dict{Symbol, Any}()
+
+    return Sampler(alg, info, s, state)
+end
+
+# always accept in the first step
+function step!(::AbstractRNG, ::Model, spl::Sampler{<:ESS}, ::Integer; kwargs...)
+    return Transition(spl)
+end
+
+function step!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{<:ESS},
+    ::Integer,
+    ::Transition;
+    kwargs...
+)
+    # recompute log-likelihood in logp
+    vi = spl.state.vi
+    if spl.selector.tag !== :default
+        runmodel!(model, vi, spl)
+    end
+
+    # obtain mean of distribution
+    vns = _getvns(vi, spl)
+    length(vns) == 1 || error("[ESS] does only support one parameter")
+    dist = getdist(vi, vns[1][1])
+    μ = vectorize(dist, mean(dist))
```

Comment:
Are there better ways to retrieve the mean of the prior distribution of the variable of interest? BTW, IMO having to write `vns[1][1]` is quite unintuitive.

Reply:
Not that I know of. @mohamed82008, do you know of a quicker way to get to a variable's distribution in a `VarInfo` outside of the `observe`/`assume` pattern?

Reply:
`Turing.Core.RandomVariables.getdist(vi, vn)`

Reply:
Oh he is already using `getdist`.